libraries/lib-module-manager/PluginHost.cpp

```diff
@@ -0,0 +1,220 @@
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  @file PluginHost.cpp
+
+  @author Vitaly Sverchinsky
+
+  Part of lib-module-manager library
+
+**********************************************************************/
+
+#include "PluginHost.h"
+
+#include <wx/app.h>
+#include <wx/log.h>
+#include <wx/module.h>
+#include <wx/process.h>
+
+#include "BasicUI.h"
+#include "PathList.h"
+#include "FileNames.h"
+#include "ModuleManager.h"
+#include "IPCClient.h"
+#include "PlatformCompatibility.h"
+#include "PluginManager.h"
+
+namespace
+{
+   //Attempts to instantiate plugin module and put plugin descriptors into result
+   void Discover(details::PluginValidationResult& result, const wxString& providerId, const wxString& pluginPath)
+   {
+      try
+      {
+         if(auto provider = ModuleManager::Get().CreateProviderInstance(providerId, wxEmptyString))
+         {
+            TranslatableString errorMessage{};
+
+            provider->DiscoverPluginsAtPath(pluginPath, errorMessage, [&result](PluginProvider *provider, ComponentInterface *ident)
+            {
+               //Workaround: use DefaultRegistrationCallback to create all descriptors for us
+               //and then put a copy into result
+               auto id = PluginManager::DefaultRegistrationCallback(provider, ident);
+               if(const auto desc = PluginManager::Get().GetPlugin(id))
+                  result.Add(PluginDescriptor { *desc });
+               return id;
+            });
+            if(!errorMessage.empty())
+               result.SetError(errorMessage.Translation());
```

Comment:
Now I'm wondering about the proper internationalization of error messages.  Does `Translation` really work, or has the host process not intiailized translations before this point is reached?  If it doesn't, maybe `Translation()` can be applied to the msgid on the host side, if the `TranslatableString` is a simple one without format arguments.

Reply:
Hm, didn't think of it, thanks. It's hard to answer such questions without a complete understanding of requirement. Though we can initialize i18n on the host side too, but then there is a question of language preference change. So, I don't have a complete answer, not use i18n in messages is a possible solution I think.